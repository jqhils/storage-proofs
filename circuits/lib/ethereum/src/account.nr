use crate::account_with_storage::Account;
use crate::header::get_header;
use crate::misc::types::{Address, Bytes32, BYTES32_LENGTH};
use crate::verifiers::account::verify_account;
use crate::serde::Serde;
use crate::merkle_patricia_proofs::proof::ProofInput;

global MAX_KEY_LEN = 32;
global MAX_PREFIXED_KEY_NIBBLE_LEN = 66; // (MAX_KEY_LEN + 1) * 2
global MAX_ACCOUNT_DEPTH_NO_LEAF_M = 10; // Emperically correct values to be determined after we scan ethereum state trie.

global MAX_ACCOUNT_STATE_LEN = 110; // Values taken from accountProofConfig in account.ts.
global MAX_ACCOUNT_LEAF_LEN = 148;

struct AccountWithinBlock {
    account: Account,
    block_hash: Bytes32,
}

impl Eq for AccountWithinBlock {
    fn eq(self, other: Self) -> bool {
        (self.account == other.account) & (self.block_hash == other.block_hash)
    }
}

type AccountWithStateProofM = (Account, ProofInput<MAX_PREFIXED_KEY_NIBBLE_LEN, MAX_ACCOUNT_STATE_LEN, MAX_ACCOUNT_DEPTH_NO_LEAF_M, MAX_ACCOUNT_LEAF_LEN>);

type ProofInputSerialized<LEN> = [Field; LEN];

pub fn get_account(chain_id: Field, block_no: u64, address: Address) -> AccountWithinBlock {
    let (account, state_proof) = get_account_unconstrained_M(chain_id, block_no, address);
    let header = get_header(chain_id, block_no);
    verify_account(address, account, state_proof, header.state_root);
    AccountWithinBlock { account, block_hash: header.hash }
}

#[oracle(get_account)]
unconstrained fn get_account_oracle<PROOF_INPUT_LEN>(
    _chain_id: Field,
    _block_no: u64,
    _address: [u8; 20]
) -> (Account, ProofInputSerialized<PROOF_INPUT_LEN>) {}

unconstrained fn get_account_unconstrained_M(chain_id: Field, block_no: u64, address: Address) -> AccountWithStateProofM {
    let (account, proof_input) = get_account_oracle(chain_id, block_no, address);
    let proof_input = Serde::deserialize(proof_input);

    (account, proof_input)
}
