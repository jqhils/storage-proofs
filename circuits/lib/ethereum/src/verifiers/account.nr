use crate::account::{Account, MAX_PREFIXED_KEY_NIBBLE_LEN, MAX_ACCOUNT_STATE_LEN, MAX_ACCOUNT_LEAF_LEN};
use crate::misc::{types::{Address, ADDRESS_LENGTH}, bytes::right_pad, fragment::Fragment};
use crate::rlp::decode::decode_list_of_small_strings;
use crate::rlp::types::RlpList;
use crate::merkle_patricia_proofs::proof::{ProofInput, verify_merkle_proof};
use crate::HASH_LEN;

use dep::std::hash::keccak256;

global ACCOUNT_FIELDS_COUNT = 4;
global NONCE_INDEX  = 0;
global BALANCE_INDEX = 1;
global STORAGE_ROOT_INDEX = 2;
global CODE_HASH_INDEX = 3;

pub(crate) fn assert_account_equals(account_rlp_left_padded: [u8; MAX_ACCOUNT_STATE_LEN], account: Account) {
    let account_rlp_right_padded = right_pad(account_rlp_left_padded).storage;
    let account_rlp_list: RlpList<ACCOUNT_FIELDS_COUNT> = decode_list_of_small_strings(Fragment::from_array(account_rlp_right_padded));
    assert(account_rlp_list.len == ACCOUNT_FIELDS_COUNT, "Invalid number of fields in account RLP");

    let account_rlp = Fragment::from_array(account_rlp_right_padded);
    account_rlp_list.get(NONCE_INDEX).assert_eq_u64("Nonce", account_rlp, account.nonce as u64);
    account_rlp_list.get(BALANCE_INDEX).assert_eq_u128("Balance", account_rlp, U128::from_integer(account.balance));

    account_rlp_list.get(STORAGE_ROOT_INDEX).assert_eq_bytes32("Storage root", account_rlp, account.storage_root);
    account_rlp_list.get(CODE_HASH_INDEX).assert_eq_bytes32("Code hash", account_rlp, account.code_hash);
}

fn assert_address_equals(address_hash: [u8; MAX_PREFIXED_KEY_NIBBLE_LEN], address: Address) {
    let address_hash_fragment = Fragment::new(MAX_PREFIXED_KEY_NIBBLE_LEN - HASH_LEN, HASH_LEN, address_hash);
    let other_address_hash_fragment = Fragment::from_array(keccak256(address, ADDRESS_LENGTH as u32));
    assert(address_hash_fragment.eq(other_address_hash_fragment), "Address mismatch");
}

pub fn verify_account<MAX_DEPTH_NO_LEAF>(
    address: Address,
    account: Account,
    state_proof_input: ProofInput<MAX_PREFIXED_KEY_NIBBLE_LEN, MAX_ACCOUNT_STATE_LEN, MAX_DEPTH_NO_LEAF, MAX_ACCOUNT_LEAF_LEN>,
    state_root: [u8; HASH_LEN]
) {
    assert_address_equals(state_proof_input.key, address);
    assert_account_equals(state_proof_input.value, account);
    verify_merkle_proof(
        state_proof_input.key,
        state_proof_input.value,
        state_root,
        state_proof_input.proof
    );
}
