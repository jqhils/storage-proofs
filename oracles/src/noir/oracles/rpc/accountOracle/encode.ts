import { ForeignCallOutput } from '@noir-lang/noir_js';
import { GetProofReturnType, Hex, fromRlp, isHex, keccak256, toRlp } from 'viem';
import { encodeField, encodeHex, encodeProof } from '../../common/encode.js';
import { padArray } from '../../../../util/array.js';
import { MAX_TRIE_NODE_LEN, ZERO_PAD_VALUE } from '../../common/const.js';
import { assert } from '../../../../util/assert.js';
import { accountProofConfig } from '../common/proofConfig/account.js';
import { storageProofConfig } from '../common/proofConfig/storage.js';
import { toHexString } from '../../../../ethereum/blockHeader.js';

const RLP_VALUE_INDEX = 1;

export function encodeAccount(ethProof: GetProofReturnType): ForeignCallOutput[] {
  const nonce = encodeField(ethProof.nonce);
  const balance = encodeField(ethProof.balance);
  const storageRoot = encodeHex(ethProof.storageHash);
  const codeHash = encodeHex(ethProof.codeHash);

  return [nonce, balance, storageRoot, codeHash];
}

export function encodeStateProof(ethProof: GetProofReturnType): ForeignCallOutput {
  const key = padArray(
    encodeHex(keccak256(ethProof.address)),
    accountProofConfig.maxPrefixedKeyNibbleLen,
    ZERO_PAD_VALUE,
    'left'
  );
  const value = padArray(encodeValue(ethProof.accountProof), accountProofConfig.maxValueLen, ZERO_PAD_VALUE, 'left');
  const nodes = encodeProof(
    ethProof.accountProof.slice(0, ethProof.accountProof.length - 1),
    (accountProofConfig.maxProofDepth - 1) * MAX_TRIE_NODE_LEN
  );
  const leaf = padArray(
    encodeHex(ethProof.accountProof[ethProof.accountProof.length - 1]),
    accountProofConfig.maxLeafLen,
    ZERO_PAD_VALUE
  );
  const depth = encodeField(ethProof.accountProof.length);

  return [...key, ...value, ...nodes, ...leaf, depth];
}

export function getValue(proof: Hex[]): Hex {
  const lastProofEntry = fromRlp(proof[proof.length - 1], 'hex');
  const value = lastProofEntry[RLP_VALUE_INDEX];
  assert(isHex(value), 'value should be of type Hex');
  return value;
}

export function encodeValue(proof: Hex[]): string[] {
  return padArray(encodeHex(getValue(proof)), accountProofConfig.maxValueLen, ZERO_PAD_VALUE, 'left');
}

type StorageProof = GetProofReturnType['storageProof'][number];

export function encodeStorageProof(storageKey: Hex, storageProof: StorageProof): ForeignCallOutput {
  const key = padArray(
    encodeHex(keccak256(storageKey)),
    storageProofConfig.maxPrefixedKeyNibbleLen,
    ZERO_PAD_VALUE,
    'left'
  );
  const value = padArray(
    encodeHex(toRlp(toHexString(storageProof.value))),
    storageProofConfig.maxValueLen,
    ZERO_PAD_VALUE,
    'left'
  );
  const nodes = encodeProof(
    storageProof.proof.slice(0, storageProof.proof.length - 1),
    (storageProofConfig.maxProofDepth - 1) * MAX_TRIE_NODE_LEN
  );
  const leaf = padArray(
    encodeHex(storageProof.proof[storageProof.proof.length - 1]),
    storageProofConfig.maxLeafLen,
    ZERO_PAD_VALUE
  );
  const depth = encodeField(storageProof.proof.length);

  return [...key, ...value, ...nodes, ...leaf, depth];
}
