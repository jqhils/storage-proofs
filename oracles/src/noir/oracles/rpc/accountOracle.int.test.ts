import { describe, expect, it } from 'vitest';
import { OFFSETS, getAccountOracle } from './accountOracle.js';
import { createMockMultiChainClient } from '../../../ethereum/mockClient.js';
import { accountProofConfig } from './common/proofConfig/account.js';

describe('accountOracle', () => {
  it('getAccountOracle', async () => {
    const mainnetChainIdInNoirFormat = '0x01';
    // prettier-ignore
    const vitalikAccountAddressInNoirFormat = [
        "0xd8", "0xda", "0x6b", "0xf2", "0x69", "0x64", "0xaf", "0x9d", "0x7e", "0xed", 
        "0x9e", "0x03", "0xe5", "0x34", "0x15", "0xd3", "0x7a", "0xa9", "0x60", "0x45"
      ]
    const londonBlockNumberInNoirFormat = '0xc5d488';
    const mockFilePaths = [
      './fixtures/mainnet/london/vitalik_balance/eth_getBlockByHash_12965000.json',
      './fixtures/mainnet/london/vitalik_balance/eth_getProof_12965000.json'
    ];
    const multiChainClient = await createMockMultiChainClient(mockFilePaths);
    const account = await getAccountOracle(multiChainClient, [
      [mainnetChainIdInNoirFormat],
      [londonBlockNumberInNoirFormat],
      vitalikAccountAddressInNoirFormat
    ]);
    expect(account[OFFSETS.NONCE]).toStrictEqual('0x02cb');
    expect(account[OFFSETS.BALANCE]).toStrictEqual('0x019c54c1cc8b1ad5994d');

    const proofInputKeyPart = account[OFFSETS.PROOF_INPUT].slice(0, accountProofConfig.maxPrefixedKeyNibbleLen);
    expect(proofInputKeyPart).toMatchSnapshot();
    const proofInputValuePart = account[OFFSETS.PROOF_INPUT].slice(
      accountProofConfig.maxPrefixedKeyNibbleLen,
      accountProofConfig.maxPrefixedKeyNibbleLen + accountProofConfig.maxValueLen
    );
    expect(proofInputValuePart).toMatchSnapshot();
    const proofInputDepthPart = account[OFFSETS.PROOF_INPUT][account[OFFSETS.PROOF_INPUT].length - 1];
    expect(proofInputDepthPart).toStrictEqual('0x09');
  });
});
