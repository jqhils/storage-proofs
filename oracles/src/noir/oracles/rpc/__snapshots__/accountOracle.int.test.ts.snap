// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`accountOracle > getAccountOracle 1`] = `
[
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x06",
  "0xe1",
  "0x20",
  "0xc2",
  "0xc3",
  "0x54",
  "0x7c",
  "0x60",
  "0xee",
  "0x47",
  "0xf7",
  "0x12",
  "0xd3",
  "0x2e",
  "0x5a",
  "0xcf",
  "0x38",
  "0xb3",
  "0x5d",
  "0x1c",
  "0xc6",
  "0x2e",
  "0x23",
  "0xb0",
  "0x55",
  "0xa6",
  "0x9b",
  "0xb8",
  "0x82",
  "0x84",
  "0xc2",
  "0x81",
]
`;

exports[`accountOracle > getAccountOracle 2`] = `
[
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0x00",
  "0xf8",
  "0x50",
  "0x82",
  "0x02",
  "0xcb",
  "0x8a",
  "0x01",
  "0x9c",
  "0x54",
  "0xc1",
  "0xcc",
  "0x8b",
  "0x1a",
  "0xd5",
  "0x99",
  "0x4d",
  "0xa0",
  "0x56",
  "0xe8",
  "0x1f",
  "0x17",
  "0x1b",
  "0xcc",
  "0x55",
  "0xa6",
  "0xff",
  "0x83",
  "0x45",
  "0xe6",
  "0x92",
  "0xc0",
  "0xf8",
  "0x6e",
  "0x5b",
  "0x48",
  "0xe0",
  "0x1b",
  "0x99",
  "0x6c",
  "0xad",
  "0xc0",
  "0x01",
  "0x62",
  "0x2f",
  "0xb5",
  "0xe3",
  "0x63",
  "0xb4",
  "0x21",
  "0xa0",
  "0xc5",
  "0xd2",
  "0x46",
  "0x01",
  "0x86",
  "0xf7",
  "0x23",
  "0x3c",
  "0x92",
  "0x7e",
  "0x7d",
  "0xb2",
  "0xdc",
  "0xc7",
  "0x03",
  "0xc0",
  "0xe5",
  "0x00",
  "0xb6",
  "0x53",
  "0xca",
  "0x82",
  "0x27",
  "0x3b",
  "0x7b",
  "0xfa",
  "0xd8",
  "0x04",
  "0x5d",
  "0x85",
  "0xa4",
  "0x70",
]
`;
